### DISCUSSION TITLE
Pydantic V2 - Define custom objects as path parameters

### DISCUSSION THREAD

**Discussion by alex-pobeditel-2004** [fastapi/fastapi#10300] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import FastAPI
from pydantic import Field


NonNegativeInt32 = Annotated[int, Field(ge=0, le=2_147_483_647)]
app = FastAPI()


@app.get("/{some_number}")
def echo(
        some_number: NonNegativeInt32
) -> NonNegativeInt32:
    return some_number
```

### Description

I try to use custom type (constrained `int`) as the route **path** parameter and execution of code above gives me such an error:
```
Traceback (most recent call last):
  File "C:\Users\user\test.py", line 13, in <module>
    def echo(                                                                                             
  File "C:\Users\user\.virtualenvs\test-vvPeaeos\lib\site-packages\fastapi\routing.py", line 704, in decorator
    self.add_api_route(
  File "C:\Users\user\.virtualenvs\test-vvPeaeos\lib\site-packages\fastapi\routing.py", line 643, in add_api_route
    route = route_class(
  File "C:\Users\user\.virtualenvs\test-vvPeaeos\lib\site-packages\fastapi\routing.py", line 489, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
  File "C:\Users\user\.virtualenvs\test-vvPeaeos\lib\site-packages\fastapi\dependencies\utils.py", line 261, in get_dependant
    type_annotation, depends, param_field = analyze_param(
  File "C:\Users\user\.virtualenvs\test-vvPeaeos\lib\site-packages\fastapi\dependencies\utils.py", line 410, in analyze_param
    assert isinstance(field_info, params.Path), (
AssertionError: Cannot use `FieldInfo` for path param 'some_number'
```
Such custom class worked with Pydantic V1 (then it was a subclass of ConstrainedInt, but here is no more ConstrainedInt so I tried to move to `Annotated`).

Is there is a proper way to use such custom classes in path parameters?

### Operating System

Linux, Windows

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.10.1

### Additional Context

I've found description of similar problems but they are related to **query parameters**. It looks like these problems have common roots.

Issue with query parameters problem: https://github.com/tiangolo/fastapi/issues/10259
Pull request with query parameters problem fix: https://github.com/tiangolo/fastapi/pull/10109 (this fix does not help with problem described above)

**Comment by alex-pobeditel-2004** [comment]
@Kludex sorry for direct mention, but I think this can be a real issue if there was no intention to change this FastAPI behavior with upgrade to Pydantic 2.X.
Could you please have a quick look at the provided example and say if such usage of types in endpoint definition is acceptable?

**Reply by lietu** [reply]
So I managed to set up more custom validations for path and query parameters like this: https://stackoverflow.com/a/77200692/3989287

Maybe it helps with this one as well. I would hope that FastAPI would add a proper support for passing information from `Annotated` on the parameters defined on the function argument level too, so it would get a little easier to define more comprehensive validations.

**Reply by alex-pobeditel-2004** [reply]
I've also found another discussion - there is a point that `Field` can't be used for path parameters now and it's possible to define similar validation class, but with `Path`:  
https://github.com/tiangolo/fastapi/discussions/10105#discussioncomment-6779426  
Didn't try it myself yet but I think that should solve the problem, at least for now.

**Comment by YuriiMotov** [comment, marked as answer]
Fixed.
Works well in current version (fastapi-0.115.12, pydantic-2.11.5)
